src/Workspaces/Core/Portable/SourceGeneratorTelemetry/SourceGeneratorTelemetryCollectorWorkspaceService.cs

```diff
@@ -0,0 +1,78 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
+using System.Text;
+using Microsoft.CodeAnalysis.Internal.Log;
+
+namespace Microsoft.CodeAnalysis.SourceGeneratorTelemetry
+{
+    internal class SourceGeneratorTelemetryCollectorWorkspaceService : ISourceGeneratorTelemetryCollectorWorkspaceService
+    {
+        private record GeneratorTelemetryKey
+        {
+            public GeneratorTelemetryKey(ISourceGenerator generator)
+            {
+                Identity = new SourceGeneratorIdentity(generator);
+                FileVersion = FileVersionInfo.GetVersionInfo(generator.GetGeneratorType().Assembly.Location).FileVersion ?? "(null)";
+            }
+
+            // TODO: mark these 'required' when we have the attributes in place
+            public SourceGeneratorIdentity Identity { get; init; }
+            public string FileVersion { get; init; }
+        }
+
+        /// <summary>
+        /// Cache of the <see cref="GeneratorTelemetryKey"/> for a generator to avoid repeatedly reading version information from disk;
+        /// this is a ConditionalWeakTable so having telemetry for older runs doesn't keep the generator itself alive.
+        /// </summary>
+        private readonly ConditionalWeakTable<ISourceGenerator, GeneratorTelemetryKey> _generatorTelemetryKeys = new ConditionalWeakTable<ISourceGenerator, GeneratorTelemetryKey>();
+
+        private readonly StatisticLogAggregator<GeneratorTelemetryKey> _elapsedTimeByGenerator = new StatisticLogAggregator<GeneratorTelemetryKey>();
+        private readonly StatisticLogAggregator<GeneratorTelemetryKey> _producedFilesByGenerator = new StatisticLogAggregator<GeneratorTelemetryKey>();
+
+        private GeneratorTelemetryKey GetTelemetryKey(ISourceGenerator generator) => _generatorTelemetryKeys.GetValue(generator, static g => new GeneratorTelemetryKey(g));
+
+        public void CollectRunResult(GeneratorDriverRunResult driverRunResult, GeneratorDriverTimingInfo driverTimingInfo)
+        {
+            foreach (var generatorTime in driverTimingInfo.GeneratorTimes)
+            {
+                _elapsedTimeByGenerator.AddDataPoint(GetTelemetryKey(generatorTime.Generator), generatorTime.ElapsedTime);
+            }
+
+            foreach (var generatorResult in driverRunResult.Results)
+            {
+                _producedFilesByGenerator.AddDataPoint(GetTelemetryKey(generatorResult.Generator), generatorResult.GeneratedSources.Length);
+            }
+        }
+
+        public void ReportStatisticsAndClear(FunctionId functionId)
+        {
+            foreach (var pair in _elapsedTimeByGenerator)
```

Comment:
```suggestion
            foreach (var (telemetryKey, elapsedTimeCounter) in _elapsedTimeByGenerator)
```

Reply:
That didn't work since the type didn't have a Deconstruct method, but I guess I forgot to check if we have an extension method somewhere.

Reply:
Yup.  there's a deconstruct available :)

Reply:
Then we either have a bug or a missing feature where we should trigger add using in that case...